enhancements/authentication/login-logout-events.md

```diff
@@ -0,0 +1,344 @@
+
+```json
+{
+  "kind": "Event",
+  "apiVersion": "audit.k8s.io/v1",
+  "level": "RequestResponse",
+...
+  "verb": "delete",
+...
+  "responseObject": {
+    "kind": "Status",
+    "apiVersion": "v1",
+    "metadata": {},
+    "status": "Success",
+    "details": {
+      "name": "sha256~sometoken",
+      "group": "oauth.openshift.io",
+      "kind": "oauthaccesstokens",
+      "uid": "444845f0-1348-4ab1-95be-d0644a6bd11c"
+    }
+  },
+...
+}
+```
+
+To log the complete object we have to set `ReturnDeletedObject: true` on the oauth-apiserver REST Store: https://github.com/openshift/oauth-apiserver/blob/6e0f92194d5a25728c826fefb2d99c5e88ebb5e5/pkg/oauth/apiserver/registry/oauthaccesstoken/etcd/etcd.go#L30-L46.
+
+This will allow to introspect the deleted user:
+
+```json
+{
+   "kind": "Event",
+   "apiVersion": "audit.k8s.io/v1",
+   "level": "RequestResponse",
+...
+    "verb": "delete",
+...
+    "responseObject": {
+      "kind": "OAuthAccessToken",
+      "apiVersion": "oauth.openshift.io/v1",
+      "metadata": {
+...
+        "userName": "exampleuser",
+      },
+      "userUID": "123",
+      "authorizeToken": "sha256~sometoken"
+    },
+...
+  }
+```
+
+### Event construction
+
+In all cases above an [Info](https://github.com/openshift/oauth-server/blob/961295a3151da3fff9daa5ea0ab8c0bf92a7d7e6/vendor/k8s.io/apiserver/pkg/authentication/user/user.go#L20) entity is available with user information which can be used to fill audit event data.
+
+### Integration in must-gather
+
+The existing `oc adm must-gather -- /usr/bin/gather_audit_logs` tool will also fetch oauth-server audit logs.
+
+### Graduation Criteria
+
+This feature is planned for OpenShift 4.10.
+
+#### Dev Preview -> Tech Preview
+
+N/A
+
+#### Tech Preview -> GA
+
+N/A
+
+#### Removing a deprecated feature
+
+N/A
+
+### Upgrade / Downgrade Strategy
+
+As we are introducing a new property with default values, there are not expected risks for upgrading.
+
+Downgrading would not cause any data loss or inconsistencies. The oauth-apiserver will just prune the fields in the request. The data itself will stay in etcd and will not be rewritten.
+
+### Version Skew Strategy
+
+All components will be developed with the concern of version skew in mind. So they will handle the missing of the `Audit` property.
+
+### Test Plan
+
+Ideally we test our feature with:
+
+- unit tests,
+- e2e tests in origin and
+- tests at must-gather.
+
+#### E2E testing steps
+
+In order to test the audit logging as mentioned in the porposal:
+
+**Verify audit logging for login**
+
+1. Create a cluster with OpenShift installed on it.
+2. Verify that the version deployed contains this enhancement.
+3. Verify that the custom resource `oauths.config.openshift.io` exists and contains the properties:
+
+```yaml
+audit:
+    profile: WriteLoginEvents
+```
+
+4. Create Login failures and Login.
+5. Execute `oc adm must-gather`.
```

Comment:
No, not like that (ditto for all the must-gathers below).

https://docs.openshift.com/container-platform/4.9/security/audit-log-view.html#nodes-nodes-audit-log-basic-viewing_audit-log-view

Reply:
Whats wrong with must gather?

Reply:
You're proposing to create a privileged container in the cluster to pull data from the cluster to get a tar that you need to process instead of doing a streamed API call

Reply:
Sure, for automated things :+1: :+1: :heart_eyes: